spinalcordtoolbox/image.py

```diff
@@ -279,7 +282,22 @@ def __init__(self, param=None, hdr=None, orientation=None, absolutepath=None, di
 
         # load an image from file
         if isinstance(param, str):
-            self.loadFromPath(param, verbose)
+            try:
+                self.loadFromPath(param, mmap, verbose)
+            except OSError as e:
+                if str(e) == "[Errno 24] Too many open files":
+                    try:
+                        # Only load images into memory if there are no more available file descriptors
+                        self.loadFromPath(param, False, verbose)
+                    except OSError as e:
+                        # If for some reason, an OSError is still thrown, then give the user a helpful error message
+                        if str(e) == "[Errno 24] Too many open files":
+                            raise OSError("[Errno 24] Too many open files. Please try increasing your system's file "
```

Comment:
Another nitpick: since [`OSError` has a bunch of other useful fields](https://docs.python.org/3/library/exceptions.html#OSError), it would be nice to preserve them.

One way would be to use second constructor, which accepts the other fields (but `winerror` only exists on Windows, so that gets complicated).

Another way, which is maybe a little gross, but seems appropriate to me, would be to modify `e` directly before re-raising it:
```py3
except OSError as e:
    if e.errno == errno.EMFILE:
        e.strerror += '. Please try...'
    raise  # with no argument, this re-raises the original exception
```

Reply:
Oh gosh, I had no idea `OSError` had attributes! Thank you for pointing this out to me. :slightly_smiling_face:

Reply:
Works perfectly here, too!

Reply:
> Another way, which is maybe a little gross, but seems appropriate to me, would be to modify e directly before re-raising it

Just out of curiosity, which part of this feels gross to you? Or, phrased another way, which aspect of this proposed change is giving you pause?

Reply:
I guess it's an instinctive reaction to mutating the innards of a object that we didn't create, but actually it's a documented public field of `OSError`, so it's really fine. And I can't imagine that a human-facing text attribute like this would have any guarantees that other code relies on. So it's pythonic.

Another option I thought of (which is almost certainly overkill), especially if we want to handle a specific situation without giving up any information along the way, would be to use a custom Exception subclass (something like `class MyException(Exception): pass`) and then to `raise MyException("some message") from e`, using the `raise ... from ...` form of [the `raise` statment](https://docs.python.org/3/reference/simple_stmts.html#the-raise-statement). Then, the custom exception is easy to catch (because it can't be some other kind of `OSError`), but it still contains the original `OSError` object in its `__cause__` field, and they both get displayed in a sensible way if uncaught.

Actually after testing this out with a toy example, I see that a `raise` statement inside an `except:` handler already saves the original exception object, but in the `__context__` field instead of `__cause__`. Both get displayed if uncaught, but with a slightly different text about the relationship between the two exceptions.

Reply:
> Actually after testing this out with a toy example, I see that a `raise` statement inside an `except:` handler already saves the original exception object, but in the `__context__` field instead of `__cause__`. Both get displayed if uncaught, but with a slightly different text about the relationship between the two exceptions.

This makes perfect sense! `__context__` must be tied to the "During handling of the above exception, another exception occurred" text I sometimes see during tracebacks, in which both the parent and child tracebacks are displayed.

In that case, I think I agree even more that we should modify the exception in place:

* If our goal was to display an entirely different error message describing an entirely different contextual error (e.g. "Too many motion-corrected volumes were opened during motion correction"), then I think raising a new exception would make sense?
* But since we're just trying to amend an existing error with useful advice, it feels a bit clunky to display what is basically the same traceback + error message twice, with the second error message being identical except for a bit of extra text.